Closes #54507 - Adds themable window border

eamodio opened the pull request:
Closes #54507 

@bpasero Here is the first pass at adding a themable window border. My open questions are:

- Is the split between `workbench.ts` and `layout.ts` correct? Or should I move more into `layout.ts`
- Is it OK to switch to use `vh/vw` units for sizing, rather than the manual pixel sizing we currently have?

Thoughts?

eamodio commented on src/vs/workbench/browser/workbench.ts:
```diff
@@ -462,4 +524,9 @@ export class Workbench extends Layout {
+
+	getClientArea(): Dimension {
+		const dim = super.getClientArea();
+		return this.windowBorder ? new Dimension(dim.width - 2, dim.height - 2) : dim;
```
Can't say I love this, but...

eamodio commented:
You can see the border in action by adding the following to your user `settings.json`

```
    "workbench.colorCustomizations": {
        "window.activeBorder": "#00ff00",
        "window.inactiveBorder": "#ff0000"
    }
```

AlexFiliakov commented:
This would be super useful for me, as I have a 4k monitor and often spawn multiple VSCode sessions to move between related projects. The windows often blend together, which is a nuisance.

For example, here is one VSCode window on top of another, but it's hard to tell where one ends and the other begins:
![Overlay VSCode Screenshot](https://i.imgur.com/HwhRVMJ.png)

eamodio commented:
@bpasero I moved the logic into `layout.ts` -- its not clear to me what could/should be moved into `window.ts`. I also added a first pass at a border radius on a mac, so hopefully that will help (I don't have a mac yet to test it on)

bpasero commented on src/vs/workbench/common/theme.ts:
```diff
@@ -589,6 +589,18 @@ export const NOTIFICATIONS_INFO_ICON_FOREGROUND = registerColor('notificationsIn
 	hc: editorInfoForeground
 }, nls.localize('notificationsInfoIconForeground', "The color used for the notification info icon."));
 
+export const WINDOW_ACTIVE_BORDER = registerColor('window.activeBorder', {
```
@eamodio unfortunately this means the color will appear for web users in their settings, so at least maybe add a documentation that this is not supported in web?

eamodio replied:
Will do

bpasero commented on src/vs/workbench/browser/media/style.css:
```diff
@@ -51,6 +51,23 @@ body.web {
 	position: relative;
 	z-index: 1;
 	overflow: hidden;
+	height: 100vh;
```
@eamodio this change is not clear to me and I am not sure about the consequences. Is it really needed and why this format and not `100%`?

eamodio replied:
By using `vh/wh` we can avoid having to manually size the container and avoids other issues with using % where all the parents need to be 100%, etc. So 100vh/wh will always size the container to the viewport height/width.

bpasero replied:
@eamodio still, we used to set `position: relative` programmatically, no?

eamodio replied:
We did, but it is set on `monaco-workbench` -- is there a need for the inline-style?

bpasero replied:
True, ignore.

bpasero commented on src/vs/workbench/browser/media/style.css:
```diff
@@ -51,6 +51,23 @@ body.web {
+	border-radius: 5px;
+}
+
+.monaco-workbench.border.windows {
```
@eamodio why is this rule needed?

eamodio replied:
It isn't and can be removed.

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -257,6 +264,12 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		}
+
+		// Theme changes
+		this._register(this.themeService.onThemeChange(() => this.setWindowBorder()));
```
@eamodio I would try to align with what our `Themable` class does (which unfortunately we cannot extend here) and introduce a method `updateStyles()` which then calls `this.setWindowBorder()`.

eamodio replied:
Will do

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -826,12 +892,14 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		if (!this.disposed) {
-			this._dimension = getClientArea(this.parent);
-
-			position(this.container, 0, 0, 0, 0, 'relative');
```
@eamodio why is this removed?

eamodio replied:
By using `vh/wh` in the CSS we don't need to manually size it any more

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -257,6 +264,12 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		}
+
+		// Theme changes
+		this._register(this.themeService.onThemeChange(theme => this.onThemeChanged(theme)));
```
```suggestion
		this._register(this.themeService.onThemeChange(theme => this.updateStyles()));
```

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -297,6 +310,19 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 
+	private onThemeChanged(theme: ITheme) {
+		this.updateStyles();
+	}
```
Remove

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		this.layout();
 	}
 
+	private setWindowBorder(skipLayout: boolean = false) {
```
Suggest: `updateWindowBorder`

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
+		if (windowBorder) {
+			addClass(this.container, Classes.WINDOW_BORDER);
+		}
+		else {
```
Our coding style is to put the `else` statement on the same line. Even better, I suggest to use `toggleClass` helper to remove the if/else.

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
+			windowBorder = true;
+
+			let borderColor = this.state.hasFocus ? activeBorder : inactiveBorder;
+			if (!borderColor) {
```
I do not really understand this fallback to the title color. When I tested this on mac I think I saw a white border around my window. Why do we need this?

eamodio replied:
I added that fallback, because if there is only 1 color defined (active or inactive), then when the window is focused/unfocused the border would jump on and off.

bpasero replied:
@eamodio I understand, but in that case wouldn't it be better to use the one color that is defined and not some random other color or even transparent?

eamodio replied:
Yeah, the new code uses the 1 color that is defined

bpasero commented on src/vs/workbench/browser/layout.ts:
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
+			if (!borderColor) {
+				borderColor = theme.getColor(this.state.hasFocus ? TITLE_BAR_ACTIVE_BACKGROUND : TITLE_BAR_INACTIVE_BACKGROUND);
+			}
+			this.container.style.setProperty('--window-border-color', borderColor ? borderColor.toString() : 'transparent');
```
All in all I think I would really not do any border business if no color is defined.

eamodio replied:
I removed the fallback to the titlebar color, and instead if one color is missing, it will just fallback to the other one.

bpasero commented:
As requested, looks on macOS:

![image](https://user-images.githubusercontent.com/900690/67365235-84d26000-f571-11e9-8065-a7f69c54e13b.png)
